### PR TITLE
fix: Executors and Placeholders are not working properly

### LINKED ISSUE #485: Executors and Placeholders refactoring
What do you think?

### PR DESCRIPTION
**Pull request by Ioloolo**
## 📑 description
**Some Executors nor Placeholders (the "`enchantments`") are old, not working properly, and improvements are evident.** Recent server implementations have added useful and light APIs, and Java also has. This PR closes #485 and resolves our needs.

## 🌟 Changes
* Added missing license header
* Added missing libraries or dependencies are used in enchantments
* Fixed inconsistent variable names that do not fit the code convention (e.g `p` to `player`)
* Aggressive use of `validation` feature

### Executors
- [x] ACTIONBAR
- [x] BROADCAST
- [x] BURN
- [x] CLEARCHAT
- [x] CLEARENTITY
- [x] CLEARPOTION
- [x] CLOSEGUI
- [x] CMD
- [x] CMDCON
- [x] DOORCLOSE
- [x] DOOROPEN
- [x] DOORTOGGLE
- [x] DROPITEM
- [x] EXPLOSION
- [x] FALLINGBLOCK
- [x] GIVE
- [x] GUI
- [x] ITEMFRAMEROTATE
- [x] ITEMFRAMESET
- [x] KICK
- [x] KILL
- [x] LEVEROFF
- [x] LEVERON
- [x] LEVERTOGGLE
- [x] LIGHTNING
- [x] LOG
- [x] MESSAGE
- [x] MODIFYHELDITEM
- [x] ~~MODIFYPLAYER (REMOVED)~~
- [x] MONEY
- [x] MYSQL
- [x] PERMISSION
- [x] POTION
- [x] PUSH
- [x] ROTATEBLOCK
- [x] SCOREBOARD
- [x] SERVER
- [x] SETBLOCK
- [x] SETCOUNT
- [x] SETFLYMODE
- [x] SETFLYSPEED
- [x] SETFOOD
- [x] SETGAMEMODE
- [x] SETHEALTH
- [x] SETHELDITEM
- [x] SETITEMLORE
- [x] SETITEMNAME
- [x] SETMAXHEALTH
- [x] SETOFFHAND
- [x] SETPLAYERINV
- [x] SETSATURATION
- [x] SETSLOT
- [x] SETTYPE
- [x] SETWALKSPEED
- [x] SETXP
- [x] SIGNEDIT
- [x] SOUND
- [x] SOUNDALL
- [x] SPAWN
- [x] TIME
- [x] TP
- [x] TPPOS
- [x] VELOCITY
- [x] WEATHER
### Placeholders
- [x] air
- [x] biome
- [x] blockname
- [x] cmdline
- [x] count
- [x] currenttimeseconds
- [x] emptyslot
- [x] emptyslots
- [x] entityname
- [x] exp
- [x] explevel
- [x] firstgroup
- [x] food
- [x] gamemode
- [x] group
- [x] haseffect
- [x] haspermission
- [x] health
- [x] helditem
- [x] helditemdisplayname
- [x] helditemhasenchant
- [x] helditemid
- [x] helditemlore
- [x] helditemname
- [x] id
- [x] idname
- [x] ip
- [x] isburning
- [x] isflying
- [x] isnumber
- [x] isop
- [x] issneaking
- [x] issprinting
- [x] killername
- [x] lore
- [x] maxhealth
- [x] money
- [x] mysql
- [x] name
- [x] offhanditem
- [x] onlineplayers
- [x] packlist
- [x] pitch
- [x] playerinv
- [x] playerloc
- [x] playerlocexact
- [x] playername
- [x] playeruuid
- [x] prefix
- [x] random
- [x] round
- [x] slot
- [x] suffix
- [x] time
- [x] tps
- [x] world
- [x] worldname
- [x] x
- [x] y
- [x] yaw
- [x] z

### CODE REVIEW

File: bukkit/src/main/resources/Executor/ACTIONBAR.js [hunk 1]
```diff
@@ -1,38 +1,31 @@
...
- *     (at your option) any later version.
- *
- *     This program is distributed in the hope that it will be useful,
- *     but WITHOUT ANY WARRANTY; without even the implied warranty of
- *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- *     GNU General Public License for more details.
- *
- *     You should have received a copy of the GNU General Public License
- *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *******************************************************************************/
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/ACTIONBAR.js:16]: License header for each files represent authors credit. Should restore those.

File: bukkit/src/main/resources/Executor/BROADCAST.js [hunk 2]
```diff
@@ -1,45 +1,17 @@
...
- *     (at your option) any later version.
- *
- *     This program is distributed in the hope that it will be useful,
- *     but WITHOUT ANY WARRANTY; without even the implied warranty of
- *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- *     GNU General Public License for more details.
- *
- *     You should have received a copy of the GNU General Public License
- *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *******************************************************************************/
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/BROADCAST.js:16]: *ditto*

File: bukkit/src/main/resources/Executor/BURN.js [hunk 3]
```diff
@@ -1,62 +1,26 @@
...
- *     (at your option) any later version.
- *
- *     This program is distributed in the hope that it will be useful,
- *     but WITHOUT ANY WARRANTY; without even the implied warranty of
- *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- *     GNU General Public License for more details.
- *
- *     You should have received a copy of the GNU General Public License
- *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *******************************************************************************/
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/BURN.js:16]: *ditto*

File: bukkit/src/main/resources/Executor/CLEARCHAT.js [hunk 4]
```diff
@@ -1,40 +1,15 @@
...
- *     (at your option) any later version.
- *
- *     This program is distributed in the hope that it will be useful,
- *     but WITHOUT ANY WARRANTY; without even the implied warranty of
- *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- *     GNU General Public License for more details.
- *
- *     You should have received a copy of the GNU General Public License
- *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
- *******************************************************************************/
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/CLEARCHAT.js:16]: And all rest missing deleted license headers should be restored.

**Review by Sayakie — COMMENTED** [review]
Can you reflect these missing license headers?

**Comment by Ioloolo** [comment]
> Can you reflect these missing license headers?

Oh.. I missed it.
btw I need to write a new code...

**Comment by wysohn** [comment]
@Ioloolo 
Don't be shy to add yourself as the author to whichever file you contributed

If you have entered even just one character, you have the ownership (but I hardly can imagine that would be merged xD)

**Comment by Ioloolo** [comment]
sry.. missed

**Comment by wysohn** [comment]
Any update on this?

It would be awesome to include it in our next release

**Comment by Ioloolo** [comment]
> Any update on this?
> 
> It would be awesome to include it in our next release

I don't have much time for this pr because of my studies.
As far as I know, only one executor needs refactoring.

I will have some free time later in October.
So I need to find someone to finish this pr instead of me.

**Comment by wysohn** [comment]
> > Any update on this?
> > It would be awesome to include it in our next release
> 
> I don't have much time for this pr because of my studies. As far as I know, only one executor needs refactoring.
> 
> I will have some free time later in October. So I need to find someone to finish this pr instead of me.

No need to rush

Take your time and focus on studies

Just asking cause we will have to schedule releases

File: bukkit/src/main/resources/Executor/SETITEMNAME.js [hunk 5]
```diff
@@ -14,23 +15,29 @@
...
     ]
+  ]
+};
+
+function SETITEMNAME(args) {
+  var item = args[1];
+  var name = ChatColor.translateAlternateColorCodes(Char('&'), args[0]);
+
+  if (!item || item.getType().name === 'AIR')
+    throw new Error('Item is null or air.');
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/SETITEMNAME.js:36]:
```suggestion
  if (item.getType() === Material.AIR) {
    throw new Error('Item cannot be an AIR.');
  }
```

File: bukkit/src/main/resources/Executor/SETTYPE.js [hunk 6]
```diff
@@ -14,31 +15,31 @@
...
-        else
-            item.setType(Material.valueOf(typeString))
-    }
-}
\ No newline at end of file
+  var type = args[0];
+  var item = args[1];
+
+  if (!item || item.getType().name === 'AIR')
+    throw new Error('Item is null or air.');
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/SETTYPE.js:36]:
```suggestion
  if (item.getType() === Material.AIR) {
    throw new Error('Item cannot be an AIR.');
  }
```

File: bukkit/src/main/resources/Executor/SETTYPE.js [hunk 7]
```diff
@@ -14,31 +15,31 @@
...
+
+  var material = Material.valueOf(type.toUpperCase());
```
Comment (Sayakie) [inline bukkit/src/main/resources/Executor/SETTYPE.js:38]:
```suggestion
  var material = Material.getMaterial(type.toUpperCase());
```

**Review by Sayakie — CHANGES_REQUESTED** [review]

**Comment by Sayakie** [comment]
너무 많은 변경점이 있어 몇몇 실행자는 아직 리뷰가 덜 끝났습니다.

해당 PR에서 스펀지의 실행자, 자리표시가와 버킷의 실행자까지만 리팩토링 진행 후 새 PR에서 버킷의 자리표시자 리팩토링을 시작하는게 어떨까요?

cc @wysohn

File: bukkit/src/main/resources/Executor/BROADCAST.js [hunk 8]
```diff
@@ -1,45 +1,41 @@
...
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *******************************************************************************/
+
+var Bukkit = Java.type('org.bukkit.Bukkit');
+var ChatColor = Java.type('org.bukkit.ChatColor');
+
+var validation = {
+  overloads: [[{type: 'string', name: 'message'}]],
```
[the remark this answers is not in the document]
Comment (Ioloolo) [reply bukkit/src/main/resources/Executor/BROADCAST.js:23]: I think so too.

**Comment by Ioloolo** [comment]
저도 한번 나누는게 낫다고 생각합니다.

**Comment by Sayakie** [comment]
Feel free to ping me when requested changes once be done. I would re-review changed things.

Note: Review available after next week as my exams should be end.

**Comment by Ioloolo** [comment]
I haven't requested a re-review yet because I haven't changed all comment.

I'm planning to re-review as soon as possible.
but I'm not sure about my schedul.

**Comment by Ioloolo** [comment]
In addition, Can you tell me what means sync argument in CMD.js?

**Comment by Sayakie** [comment]
> I'm planning to re-review as soon as possible.
> but I'm not sure about my schedul.

As wysohn said, no need to rush. Focus your schedule first :)

**Comment by Sayakie** [comment]
> In addition, Can you tell me what means sync argument in CMD.js?

See [Executor wiki](https://github.com/TriggerReactor/TriggerReactor/wiki/Executors_kr#cmd-cmd-spawn) for further information. Executor `CMD` consumes first argument as run as server thread (synchronously) if is equivalent to "true".

**Comment by Sayakie** [comment]
**Side note**: Should be update Executors wiki. `CMD` would be expected first parameter as boolean type, but there is no context what performs do thing.

cc @rkdrns4747

**Comment by Sayakie** [comment]
> In addition, Can you tell me what means sync argument in CMD.js?

Nevermind about that. It just was an outdated argument.

**Comment by wysohn** [comment]
> 너무 많은 변경점이 있어 몇몇 실행자는 아직 리뷰가 덜 끝났습니다.
> 
> 해당 PR에서 스펀지의 실행자, 자리표시가와 버킷의 실행자까지만 리팩토링 진행 후 새 PR에서 버킷의 자리표시자 리팩토링을 시작하는게 어떨까요?
> 
> cc @wysohn
@Sayakie 
@Ioloolo 
인터뷰 준비하느라 바빠서 확인이 늦었네요 죄송합니다

음 상관은 없다고 생각하지만 스펀지 API 7 의 경우 3.3.x 부터는 지원이 끊기는지라 나누게된다면 버킷을 먼저 넣는게 맞지 않을까요?

나누게된다면 스펀지쪽 executor랑 placeholder는 3.4나 3.5에서 스펀지 API 8 을 지원할때 다시 확인해보는게 좋을거같습니다

File: bukkit/src/main/resources/Placeholder/lore.js [hunk 9]
```diff
@@ -14,31 +15,36 @@
...
+  var iter = lores.iterator();
+  var lore = ''
+
+  while (iter.hasNext()) {
+    var itemLore = iter.next();
+
+    lore += itemLore + '\n';
+  }
+
+  lore = lore.substring(0, lore.length-1);
```
Comment (Sayakie) [inline bukkit/src/main/resources/Placeholder/lore.js:47]:
```suggestion
  var lore = JavaString.join('\n', lores);
```

File: bukkit/src/main/resources/Placeholder/prefix.js [hunk 10]
```diff
@@ -1,23 +1,27 @@
...
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *******************************************************************************/
+
+function prefix(args) {
+  if (!player)
+    throw new Error('Player is null.');
+
+  if (!vault)
+    throw new Error('This server doesn\'t has vault.')
```
Comment (Sayakie) [inline bukkit/src/main/resources/Placeholder/prefix.js:24]:
```suggestion
    throw new Error('Server has no Value plugin.')
```

File: bukkit/src/main/resources/Placeholder/suffix.js [hunk 11]
```diff
@@ -1,23 +1,27 @@
...
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *******************************************************************************/
+
+function prefix(args) {
+  if (!player)
+    throw new Error('Player is null.');
+
+  if (!vault)
+    throw new Error('This server doesn\'t has vault.')
```
Comment (Sayakie) [inline bukkit/src/main/resources/Placeholder/suffix.js:24]:
```suggestion
    throw new Error('Server has no Value plugin.')
```

**Review by Sayakie — CHANGES_REQUESTED** [review]
The review is not done yet. Will watch it again tomorrow.

**Comment by Sayakie** [comment]
Big sorry for late reply. Looks good to me, and seems there is no errorprone syntax. We have to testing whole of placeholders and executors on latest and legacy game server both.

+ Need rebase.

**Review by Sayakie — APPROVED** [review]

**Comment by Ioloolo** [comment]
My schedule is all over.
I will test the in-game environment about next week

**Review by wysohn — APPROVED** [review]
Thank you for the hard works!

These are a lot of changes, so I think the best option to test it for now is do so in the dev branch.

Lets see how it goes

**Pull request merged**